### PR TITLE
introduce simplex meshes for ExaDG::Structure

### LINKED ISSUE #307: poisson/sine does not seem to work for simplex
The application poisson/sine does not work if I simply change `ElementType::Hypercube` to `ElementType::Simplex`. @necioglu could you provide the application in a way that all the other necessary parameters are adjusted if I change to simplex meshes?

### PR DESCRIPTION
**Pull request by nfehn**
in analogy to what we did for ExaDG::Poisson ...

@necioglu @peterrum FYI

closes #307

### CODE REVIEW

File: include/exadg/structure/spatial_discretization/operator.cpp [hunk 1]
```diff
@@ -251,7 +260,13 @@ Operator<dim, Number>::fill_matrix_free_data(MatrixFreeData<dim, Number> & matri
...
 
   // dealii::Quadrature
-  matrix_free_data.insert_quadrature(dealii::QGauss<1>(param.degree + 1), get_quad_name());
+  if(this->grid->triangulation->all_reference_cells_are_hyper_cube())
+    matrix_free_data.insert_quadrature(dealii::QGauss<1>(param.degree + 1), get_quad_name());
+  else if(this->grid->triangulation->all_reference_cells_are_simplex())
+    matrix_free_data.insert_quadrature(dealii::QGaussSimplex<dim>(param.degree + 1),
+                                       get_quad_name());
+  else
+    AssertThrow(false, ExcNotImplemented());
```
Comment (necioglu) [inline include/exadg/structure/spatial_discretization/operator.cpp:269]: I want to ask a question here. Why are we using `QGauss<1>`  and not `QGauss<dim>` for the hypercube case? I had to introduce the same difference for the Poisson case.

This is also the reason why we can't use the following function: https://www.dealii.org/developer/doxygen/deal.II/classReferenceCell.html#a56bb8396fda394d712fc4f19de4b02c4

Comment (peterrum) [reply include/exadg/structure/spatial_discretization/operator.cpp:269]: > I want to ask a question here. Why are we using `QGauss<1>` and not `QGauss<dim>` for the hypercube case? I had to introduce the same difference for the Poisson case.
> 
> This is also the reason why we can't use the following function: https://www.dealii.org/developer/doxygen/deal.II/classReferenceCell.html#a56bb8396fda394d712fc4f19de4b02c4

You can simply switch to `QGauss<dim>`. `MatrixFree` will detect the tensor-product structure internally and will create the 1D quadrature for you.

**Review by peterrum — APPROVED** [review]

File: include/exadg/structure/spatial_discretization/operator.cpp (hunk 1 above)
Comment (nfehn) [reply include/exadg/structure/spatial_discretization/operator.cpp:269]: should I change it to `dim` here?

**Comment by nfehn** [comment]
before merging, I need to add an application solving a problem on a simplex mesh.

File: include/exadg/structure/spatial_discretization/operator.cpp (hunk 1 above)
Comment (peterrum) [reply include/exadg/structure/spatial_discretization/operator.cpp:269]: > should I change it to dim here?

I don't have an opinion on that. I guess the `MatrixFreeWrapper` will do the conversion anyway interanlly?

File: include/exadg/poisson/user_interface/parameters.cpp [hunk 2]
```diff
@@ -66,6 +66,13 @@ Parameters::check() const
   AssertThrow(solver != Solver::Undefined, dealii::ExcMessage("parameter must be defined."));
   AssertThrow(preconditioner != Preconditioner::Undefined,
               dealii::ExcMessage("parameter must be defined."));
+
+  if(preconditioner == Preconditioner::Multigrid and multigrid_data.use_global_coarsening == true)
+  {
+    AssertThrow(grid.create_coarse_triangulations == true,
+                dealii::ExcMessage(
+                  "This parameter needs to be true for global coarsening multigrid."));
```
[the remark this answers is not in the document]
Comment (necioglu) [reply include/exadg/poisson/user_interface/parameters.cpp:74]: From the issue #303, it is clear that the existing ExcMesage, created by me, is not clear enough.

[the remark this answers is not in the document]
Comment (nfehn) [reply include/exadg/poisson/user_interface/parameters.cpp:74]: I agree. I'll have a look at this in the present PR.

File: include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp [hunk 3]
```diff
@@ -638,7 +641,9 @@ MultigridPreconditionerBase<dim, Number>::initialize_affine_constraints(
 
   // collect all boundary functions and translate to format understood by
   // deal.II to cover all boundaries at once
-  dealii::Functions::ZeroFunction<dim, MultigridNumber> zero_function;
+  dealii::Functions::ZeroFunction<dim, MultigridNumber> zero_function(
+    dof_handler.get_fe().n_components());
+  ;
```
Comment (necioglu) [inline include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp:646]: There is an extra `;` here.

Comment (nfehn) [reply include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp:646]: ah. Do you want to fix this in a separate PR?

Comment (necioglu) [reply include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp:646]: will do.

**Pull request merged**